src/Microsoft.DotNet.CodeFormatting/Rules/HasNewLineAtEndOfFileFormattingRule.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.DotNet.CodeFormatting.Rules
+{
+    [SyntaxRuleOrder(SyntaxRuleOrder.HasNewLineAtEndOfFile)]
+    internal sealed class HasNewLineAtEndOfFileFormattingRule : ISyntaxFormattingRule
+    {
+        public SyntaxNode Process(SyntaxNode syntaxRoot)
+        {
+            var node = syntaxRoot.DescendantNodes().LastOrDefault();
+
+            if (node != null)
+            {
+                syntaxRoot = syntaxRoot.ReplaceNode(node, RemoveNewLines(node));
+            }
+
+            var token = syntaxRoot.DescendantTokens().Single(x => x.IsKind(SyntaxKind.EndOfFileToken));
+
+            return syntaxRoot.ReplaceToken(token, AdjustNewLines(token));
+        }
+
+        private static SyntaxNode RemoveNewLines(SyntaxNode node)
+        {
+            var newTrivia = Enumerable.Empty<SyntaxTrivia>();
+
+            if (node.HasTrailingTrivia)
+            {
+                newTrivia = node.GetTrailingTrivia().Where(x => !x.IsKind(SyntaxKind.EndOfLineTrivia));
+            }
+
+            return node.WithTrailingTrivia(newTrivia);
+        }
+
+        private static SyntaxToken AdjustNewLines(SyntaxToken token)
+        {
+            var newTrivia = Enumerable.Empty<SyntaxTrivia>();
+
+            if (token.HasLeadingTrivia)
+            {
+                newTrivia = token.LeadingTrivia.Where(x => !x.IsKind(SyntaxKind.EndOfLineTrivia));
```

Comment:
This needs to be a bit more nuanced here.  Consider the following code:

``` csharp
class C
{

}
// comment 1
// comment 2 
```

The result of running this rule would be 

``` csharp
class C
{

}
// comment 1 // comment 2
``

This is because the leading trivia can include comments (both structured and unstructured).  

I had to make a similar change for braces above a new line.  It has virtually the same problems as this particular change.  May want to just make this a helper method. 

https://github.com/dotnet/codeformatter/blob/master/src/Microsoft.DotNet.CodeFormatting/Rules/BraceNewLineRule.cs#L142